src/integer_mod_q/modulus_polynomial_ring_zq/get.rs

```diff
@@ -8,11 +8,120 @@
 
 //! Implementations to get information about a [`ModulusPolynomialRingZq].
 
-use crate::{integer::Z, integer_mod_q::ModulusPolynomialRingZq};
+use crate::{
+    error::MathError,
+    integer::Z,
+    integer_mod_q::{Modulus, ModulusPolynomialRingZq, Zq},
+    traits::GetCoefficient,
+    utils::index::evaluate_index,
+};
 use flint_sys::{
     fmpz::fmpz_set,
+    fmpz_mod_poly::fmpz_mod_poly_get_coeff_fmpz,
     fq::{fq_ctx_degree, fq_ctx_struct},
 };
+use std::{fmt::Display, rc::Rc};
+
+impl GetCoefficient<Zq> for ModulusPolynomialRingZq {
+    /// Returns the coefficient of a polynomial [`ModulusPolynomialRingZq`] as a [`Zq`].
+    ///
+    /// If an index is provided which exceeds the highest set coefficient, `0` is returned.
+    ///
+    /// Parameters:
+    /// - `index`: the index of the coefficient to get (has to be positive)
+    ///
+    /// Returns the coefficient as a [`Zq`], or a [`MathError`] if the provided index
+    /// is negative and therefore invalid, or it does not fit into an [`i64`].
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::traits::*;
+    /// use qfall_math::integer_mod_q::{Zq, ModulusPolynomialRingZq};
+    /// use std::str::FromStr;
+    ///
+    /// let poly = ModulusPolynomialRingZq::from_str("4  0 1 2 3 mod 17").unwrap();
+    ///
+    /// let coeff_0: Zq = poly.get_coeff(0).unwrap();
+    /// let coeff_1: Zq = poly.get_coeff(1).unwrap();
+    /// let coeff_4: Zq = poly.get_coeff(4).unwrap();
+    ///
+    /// assert_eq!(Zq::from((0, 17)), coeff_0);
+    /// assert_eq!(Zq::from((1, 17)), coeff_1);
+    /// assert_eq!(Zq::from((0, 17)), coeff_4);
+    /// ```
+    ///
+    /// # Errors and Failures
+    /// - Returns a [`MathError`] of type [`OutOfBounds`](MathError::OutOfBounds) if
+    ///     either the index is negative or it does not fit into an [`i64`].
+    fn get_coeff(&self, index: impl TryInto<i64> + Display) -> Result<Zq, MathError> {
+        let index = evaluate_index(index)?;
+        let mut out_z = Z::default();
+
+        unsafe {
+            fmpz_mod_poly_get_coeff_fmpz(
+                &mut out_z.value,
+                &self.modulus.modulus[0],
+                index,
+                &self.get_fq_ctx_struct().ctxp[0],
+            )
+        }
+
+        let modulus = Modulus {
+            modulus: Rc::new(self.get_fq_ctx_struct().ctxp[0]),
+        };
```

Comment:
I think you introduce a memory leak here:

In modulus you have saved the value with a reference counter, but still point towards the value saved in the polynomial. Meaning: as long as the polynomial is around, the modulus will only be dropped if all instantiations of the modulus are destroyed. But if this happens, the polynomial is still there, but the referenced modulus has been dropped. The other way around also works the same way

Reply:
I think that self.get_fq_ctx_struct().ctxp[0] is copied here (Since fmpz_mod_ctx has the Copy trait) and thus, no leak occurs

Reply:
I see... but their copy trait implementation is wrong... It is derived, but it has fmpz inside of it, which are not properly copied, meaning that this object causes memory leaks

Reply:
OK, that is bad. I resolved it now by creating a new instance based on the reference to the ctx object